src/python/pants/engine/target.py

```diff
@@ -714,6 +714,27 @@ def expect_single(self) -> Target:
         return self[0]
 
 
+# NB: ShouldTraverseDepsPredicate is defined below after SpecialCasedDependencies.
+# Implementations are defined here, however, before *DependenciesRequest types.
+
+
+def traverse_if_dependencies_field(target: Target, field: Field) -> bool:
```

Comment:
I split up the definition of `ShouldTraversedepsPredicate` from the implementations to clean up some typing oddities noted in the reviews on #19155.

Reply:
Thanks for using the more concrete types :raised_hands: 

WHy can't these methods just be defined below `ShouldTraverseDepsPredicate` below? Then you wouldn't need the `NB` comments about location.

Also, you'd be able to use https://typing-extensions.readthedocs.io/en/latest/#assert_type (if you want to. type-checkers will check the type at the callsite, but this'd check it at the declaration site)

Reply:
> Thanks for using the more concrete types 🙌

I'm missing something here..? I only see `field: Field` still..

Reply:
That results in 
```
  File "..../src/python/pants/engine/target.py", line 848, in CoarsenedTargetsRequest
    should_traverse_deps_predicate: ShouldTraverseDepsPredicate = traverse_if_dependencies_field,
NameError: name  'traverse_if_dependencies_field' is not defined
```

So, in order to use these as the default values, they have to be defined before the dataclasses that use it. We get away with putting `ShouldTraverseDepsPredicate` later because type hints support forward references (`from __future__ import annotations`).

This file is so long... I wish we could split it up somehow. And yet there are so many relationships between everything. This is just one more code smell.

Reply:
> > Thanks for using the more concrete types 🙌
> 
> I'm missing something here..? I only see `field: Field` still..

Oops. I missed updating that type annotation. I'll push a fix momentarily...

Reply:
See my comment below: https://github.com/pantsbuild/pants/pull/19272#discussion_r1223461655

Reply:
I tried using `assert_type` to validate the definition of these functions, but mypy complains that:
```
error: Expression is of type "Callable[[Target, Union[Dependencies, SpecialCasedDependencies]], bool]", not "Callable[[Target, Union[Dependencies, SpecialCasedDependencies]], bool]"  [assert-type]
```
Which makes no sense. I tried calling `assert_type` in several different places, but I always got the same error. So, I'm giving up on making that work. We at least have the type and the implementations close together now, so it should be easier to keep them in sync.